sunpy/net/attr.py

```diff
@@ -23,16 +16,76 @@
 Please note that & is evaluated first, so A & B | C is equivalent to
 (A & B) | C.
 """
+import keyword
+import warnings
+from collections import defaultdict, namedtuple
 
-from __future__ import absolute_import
+from astropy.utils.misc import isiterable
 
 from sunpy.util.multimethod import MultiMethod
 from sunpy.extern.six import iteritems
 
-# XXX: Maybe allow other normal forms.
+_ATTR_TUPLE = namedtuple("attr", "name name_long desc")
+
+__all__ = ['Attr', 'DummyAttr', 'SimpleAttr', 'AttrAnd',
+           'AttrOr', 'ValueAttr', 'and_', 'or_']
+
+
+def make_tuple():
+    return _ATTR_TUPLE([], [], [])
+
 
+class AttrMeta(type):
+    """
+    We want to enable automatic discovery via tab completion of subclasses of Attrs.
+    To do this we have to create a metaclass that redefines the methods that Python uses to normally do this.
+    This would allow for example that `attrs.Instrument` to be able to tab complete to `attrs.Instrument.aia`.
+    However, it must be done by a downloader client which know what Attrs they support.
+    """
 
-class Attr(object):
+    # The aim is to register Attrs as a namedtuple of lists
+    # So we define the namedtuple outside of AttrMeta, see above.
+    _attr_registry = defaultdict(make_tuple)
+
+    def __getattr__(self, item):
+        """
+        Our method for Attrs is to register using `type(Attr)` as keys into a dictionary.
+        _attr_registry is a dictionary with a key being `type(Attr)` and the value being the namedtuple of lists.
+        As a result we index `_attr_registry` with `[self]` which will be the `type(Attr)` to access the dictionary.
+        This will return the namedtuple that has three attributes: `name`, `name_long` and `desc`. Each of which are a list.
+        `name` will be the attribute name, `name_long` is the original name passed in and `desc` the description of the object.
+        """
+        registry = self._attr_registry[self]  # Get the revelant entries.
+        names = registry.name  # All the attribute names under that type(Attr)
+        if item in names:
+            return self(registry.name_long[names.index(item)])  # We return Attr(name_long) to create the Attr requested.
+        else:
+            raise AttributeError('This attribute is not defined, please register it.')
```

Comment:
This error should be more user focused. Users don't know what registration is. I think something like 

```
f"{item} is not a known attribute of {self}"
```